shell/platform/darwin/macos/framework/Source/FlutterEngine.mm

```diff
@@ -408,39 +464,77 @@ - (void)loadAOTData:(NSString*)assetsDir {
   }
 }
 
+- (void)registerViewController:(FlutterViewController*)controller forId:(uint64_t)viewId {
+  NSAssert(controller != nil, @"The controller must not be nil.");
+  NSAssert(![controller attached],
+           @"The incoming view controller is already attached to an engine.");
+  NSAssert([_viewControllers objectForKey:@(viewId)] == nil, @"The requested view ID is occupied.");
+  // The first view controller is guaranteed to have the fixed default view ID.
+  NSAssert([_viewControllers count] != 0 || viewId == kFlutterDefaultViewId,
+           @"The first view controller should have the fixed default view ID, but is "
+           @"given %lld.",
+           viewId);
+  [controller attachToEngine:self withId:viewId];
+  NSAssert(controller.id == viewId, @"Failed to assign view ID.");
+  [_viewControllers setObject:controller forKey:@(viewId)];
+}
+
+- (void)deregisterViewControllerForId:(uint64_t)viewId {
+  FlutterViewController* oldController = [self viewControllerForId:viewId];
+  if (oldController != nil) {
+    [oldController detachFromEngine];
+    [_viewControllers removeObjectForKey:@(viewId)];
+  }
+}
+
+- (void)shutDownIfNeeded {
+  if ([_viewControllers count] == 0 && !_allowHeadlessExecution) {
+    [self shutDownEngine];
+  }
+}
+
+- (FlutterViewController*)viewControllerForId:(uint64_t)viewId {
+  FlutterViewController* controller = [_viewControllers objectForKey:@(viewId)];
+  NSAssert(controller == nil || controller.id == viewId,
+           @"The stored controller has unexpected view ID.");
```

Comment:
As per [this comment](https://github.com/flutter/engine/pull/37976/files#r1096338601), consider:

```suggestion
  NSAssert(controller != nil && controller.id == viewId,
           @"The stored controller has unexpected view ID.");
```

Reply:
Well I thought in that comment we agreed that it's ok for this method to return null?

Reply:
Right, but from my understanding this assertion fails instead of returning `null`.